In [105]:
%reload_ext autoreload
%autoreload 2
%matplotlib inline

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

import sys
sys.path.append('..')

from helper import nn
from helper import logistic_regression as lr

In [106]:
raw_X, raw_y = nn.load_data()
print(raw_X.shape)
print(raw_y.shape)

(5000, 400)
(5000,)


# prepare data

In [107]:
# add intercept=1 for x0
X = np.insert(raw_X, 0, values=np.ones(X.shape[0]), axis=1)
X.shape

(5000, 401)

In [108]:
# y have 10 categories here. 1..10, they represent digit 0 as category 10 because matlab index start at 1
# I'll ditit 0, index 0 again
y_matrix = []

for k in range(1, 11):
    y_matrix.append((raw_y == k).astype(int))

# last one is k==10, it's digit 0, bring it to the first position
y_matrix = [y_matrix[-1]] + y_matrix[:-1]
y = np.array(y_matrix)

y.shape

(10, 5000)

# train 1 model

In [109]:
t0 = lr.logistic_regression(X, y[0])

In [110]:
print(t0.shape)
y_pred = lr.predict(X, t0)
print('Accuracy={}'.format(np.mean(y[0] == y_pred)))

(401,)
Accuracy=0.9974


Is this real......

# train k model

In [113]:
k_theta = np.array([lr.logistic_regression(X, y[k]) for k in range(10)])
print(k_theta.shape)

(10, 401)


# making prediction
* think about the shape of k_theta, now you are making $X\times\theta^T$
> $(5000, 401) \times (10, 401).T = (5000, 10)$
* after that, you run sigmoid to get probabilities and for each row, you find the highest prob as the answer

In [114]:
prob_matrix = lr.sigmoid(X @ k_theta.T)

In [115]:
np.set_printoptions(suppress=True)
prob_matrix

array([[ 0.99577468,  0.        ,  0.00053556, ...,  0.00006469,
         0.00003916,  0.00172136],
       [ 0.99834655,  0.0000001 ,  0.00005615, ...,  0.00009695,
         0.0000029 ,  0.0000849 ],
       [ 0.99139761,  0.        ,  0.00056838, ...,  0.00000655,
         0.02654992,  0.00197519],
       ..., 
       [ 0.00000068,  0.04145024,  0.00320926, ...,  0.00012707,
         0.0029766 ,  0.70763098],
       [ 0.00001844,  0.00000013,  0.00000009, ...,  0.00164766,
         0.06814744,  0.86119877],
       [ 0.028808  ,  0.        ,  0.00012977, ...,  0.36631063,
         0.00498364,  0.1482392 ]])

In [116]:
y_pred = np.argmax(prob_matrix, axis=1)

In [117]:
y_answer = raw_y.copy()
y_answer[y_answer==10] = 0

In [120]:
print('Training set accuracy = {}'.format(np.mean(y_pred == y_answer)))

Training set accuracy = 0.9446
